build.assets/Dockerfile-centos7-fips

```diff
@@ -1,149 +1,27 @@
-    cmake3 \
-    # required by libbpf
-    elfutils-libelf-devel-static \
-    git \
-    net-tools \
-    # required by boringssl
-    ninja-build \
-    # required by Teleport PAM support
-    pam-devel \
-    perl-IPC-Cmd \
-    tree \
-    # used by our Makefile
-    which \
-    zip \
-    # required by libbpf
-    zlib-static && \
-    yum clean all
-
-# As mentioned above, these packages are unsigned.
-RUN yum install --nogpgcheck -y \
-        ${DEVTOOLSET}-gcc* \
-        ${DEVTOOLSET}-make && \
-    yum clean all
-
-# Install Go.
-ARG GOLANG_VERSION
-RUN mkdir -p /opt && cd /opt && curl -fsSL https://storage.googleapis.com/golang/${GOLANG_VERSION}.linux-${BUILDARCH}.tar.gz | tar xz && \
-    mkdir -p /go/src/github.com/gravitational/teleport && \
-    chmod a+w /go && \
-    chmod a+w /var/lib && \
-    chmod a-w /
-ENV GOEXPERIMENT=boringcrypto \
-    GOPATH="/go" \
-    GOROOT="/opt/go" \
-    PATH="/opt/llvm/bin:$PATH:/opt/go/bin:/go/bin:/go/src/github.com/gravitational/teleport/build"
-
-# Install PAM module and policies for testing.
-COPY pam/ /opt/pam_teleport/
-RUN make -C /opt/pam_teleport install
-
-RUN chmod a-w /
-
-ARG RUST_VERSION
-ENV RUSTUP_HOME=/usr/local/rustup \
-     CARGO_HOME=/usr/local/cargo \
-     PATH=/usr/local/cargo/bin:$PATH \
-     RUST_VERSION=$RUST_VERSION
-
-RUN mkdir -p $RUSTUP_HOME && chmod a+w $RUSTUP_HOME && \
-    mkdir -p $CARGO_HOME/registry && chmod -R a+w $CARGO_HOME
-
-# Install Rust using the ci user, as that is the user that
-# will run builds using the Rust toolchains we install here.
-USER ci
-RUN curl --proto '=https' --tlsv1.2 -fsSL https://sh.rustup.rs | sh -s -- -y --profile minimal --default-toolchain $RUST_VERSION && \
-    rustup --version && \
-    cargo --version && \
-    rustc --version && \
-    rustup component add rustfmt clippy && \
-    rustup target add ${TARGETARCH}-unknown-linux-gnu
-
-ARG WASM_PACK_VERSION
-# Install wasm-pack for targeting WebAssembly from Rust.
-RUN cargo install wasm-pack --version ${WASM_PACK_VERSION}
-
-ARG LIBBPF_VERSION
-COPY --from=libbpf /opt/libbpf/usr /usr/libbpf-${LIBBPF_VERSION}
-
-# Download pre-built CentOS 7 assets with clang needed to build BoringSSL and BPF tools.
-COPY --from=teleport-buildbox-centos7-assets /opt/llvm /opt/llvm
-
-VOLUME ["/go/src/github.com/gravitational/teleport"]
-EXPOSE 6600 2379 2380
+ARG BUILDBOX_CENTOS7
+FROM ${BUILDBOX_CENTOS7}-${BUILDARCH}
+
+# Enable FIPS mode (BoringCrypto) for Go
+# https://github.com/golang/go/blob/master/src/crypto/internal/boring/README.md
+ENV GOEXPERIMENT=boringcrypto
+
+# Enable OpenSSL FIPS mode by default
+# https://www.openssl.org/docs/man3.0/man7/fips_module.html
+RUN echo 'config_diagnostics = 1\n\
+openssl_conf = openssl_init\n\
+\n\
+.include /usr/local/ssl/fipsmodule.cnf\n\
+\n\
+[openssl_init]\n\
+providers = provider_sect\n\
+\n\
+[provider_sect]\n\
+fips = fips_sect\n\
+base = base_sect\n\
+\n\
+[base_sect]\n\
+activate = 1\n\
+' > /usr/local/ssl/openssl.cnf
```

Comment:
Isn't this creating a file with a bunch of literal `\n`s in it?

I would expect the actual newlines within the echo string to be enough.

Reply:
It's super weird, for sure. I ran a  test with a stripped Dockerfile now and I think you are a missing a `$` in the `echo $'...'` command, like this:

```diff
- RUN echo 'config_diagnostics = 1\n\
+ RUN echo $'config_diagnostics = 1\n\
```

I also tried a heredoc and it seems to work fine, so +1 to that instead of the somewhat-esoteric echo-$-'':

```Dockerfile
RUN cat >/usr/local/ssl/openssl.cnf <<EOF
config_diagnostics = 1
openssl_conf = openssl_init

.include /usr/local/ssl/fipsmodule.cnf

[openssl_init]
providers = provider_sect

[provider_sect]
fips = fips_sect
base = base_sect

[base_sect]
activate = 1
EOF
```

(Apparently this landed in 2021 - https://www.docker.com/blog/introduction-to-heredocs-in-dockerfiles/.)

Reply:
@codingllama I remember we had some issues with heredocs in the past. @reedloden Do you remember the details? 
AFAIR we're getting `failed to solve with frontend dockerfile.v0: failed to create LLB definition:` when you build with `BUILDKIT_PROGRESS=plain` which is used by Drone AFAIR to show all logs from the Docker build. I think you also need to add `# syntax=docker/...` at the top of the file.
I'd also prefer to use them, but if we make another attempt, we need to make sure that all GHA and Drone builds are passing.

Reply:
Since we're actively moving this to GHA, let's give heredoc another try. We may have just gotten lucky with some of @camscale's recent work to make this worthwhile.

Reply:
I don't remember the problems, but thanks for giving it a shot anyway.